src/x86_64/hpet.c

```diff
@@ -117,22 +119,45 @@ static u64 hpet_main_counter()
 static void timer_config(int timer, timestamp rate, thunk t, boolean periodic)
 {
     if (!hpet_interrupts[timer]) {
-        u32 a, d;
+        u64 c = TCONF(32MODE_CNF) | TCONF(INT_ENB_CNF);
         hpet_interrupts[timer] = allocate_interrupt();
-        msi_format(&a, &d, hpet_interrupts[timer]);
-        hpet->timers[timer].fsb_int = ((u64)a << 32) | d;
+        if (hpet->timers[timer].config & TCONF(FSB_INT_DEL_CAP)) {
+            u32 a, d;
+            msi_format(&a, &d, hpet_interrupts[timer]);
+            hpet->timers[timer].fsb_int = ((u64)a << 32) | d;
+            c |= TCONF(FSB_EN_CNF);
+        } else {
+            u32 route_cap = (hpet->timers[timer].config >>
+                    HPET_TIMER_CONFIG_INT_ROUTE_CAP_SHIFT) &
+                    MASK(HPET_TIMER_CONFIG_INT_ROUTE_CAP_BITS);
+            int gsi = -1;
+            for (int i = 0; i < 64; i++) {
+                /* Avoid interrupt #2, which is the interrupt used by the I8254
+                 * timer (PIT), otherwise interrupts from the PIT will uselessly
+                 * wake up the CPU periodically. */
+                if ((i != 2) && (route_cap & (1 << i)) &&
+                        ioapic_int_is_free(i)) {
+                    gsi = i;
+                    break;
+                }
+            }
+            assert(gsi >= 0);
+            ioapic_set_int(gsi, hpet_interrupts[timer]);
+            c |= gsi << HPET_TIMER_CONFIG_INT_ROUTE_CNF_SHIFT;
+        }
         register_interrupt(hpet_interrupts[timer], t, "hpet timer");
+        hpet->timers[timer].config = c;
     }
 
-    u64 c = TCONF(FSB_EN_CNF) | TCONF(32MODE_CNF) | TCONF(INT_ENB_CNF);
     if (periodic) {
         if ((hpet->timers[timer].config & TCONF(PER_INT_CAP)) == 0) {
 	    console("HPET timer not capable of periodic interrupts.\n");
 	    return;
 	}
-	c |= TCONF(VAL_SET_CNF) | TCONF(TYPE_CNF);
+        hpet->timers[timer].config |= TCONF(VAL_SET_CNF) | TCONF(TYPE_CNF);
```

Comment:
Couldn't the two writes to this register be merged into one, as it was before?

Reply:
Before, the bits of the register that control interrupt mapping were set up statically, so we could always re-write them with fixed values, but now their contents may depend on a value (the GSI number) calculated at initialization, so any time we select between periodic and non-periodic mode we have to read first the current contents of the register. Do you prefer something like below?
```
u64 c;
if (!hpet_interrupts[timer]) {
  c = ...;
} else {
  c = hpet->timers[timer].config;
}
if (periodic) {
  ...
  c |= TCONF(VAL_SET_CNF) | TCONF(TYPE_CNF);
} else {
  c &= ~(TCONF(VAL_SET_CNF) | TCONF(TYPE_CNF));
}
hpet->timers[timer].config = c;
```

Reply:
Yes, or maybe stash a copy of the GSI? I'm assuming accesses to HPET registers cause a vm exit and are expensive, but I don't know how ESXi handles them.

Reply:
OK, I created an array of struct hpet_timer_cfg that contains both the interrupt number (which replaces the hpet_interrupts static array) and the timer configuration, so now timer_config() only does one write (and no reads) to the HPET timer config register (except for the initial interrupt configuration).